core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -281,120 +320,74 @@ public final BasicInvocationStage visitClearCommand(InvocationContext ctx, Clear
-         //for the invalidate command, we need to try to fetch the key from the data container
-         //otherwise it may be not removed
-         entryFactory.wrapEntryForWriting(ctx, key, EntryFactory.Wrap.WRAP_NON_NULL, false, true);
-         if (trace) log.tracef("Entry to be removed: %s", toStr(key));
+         // TODO: move to distribution interceptors?
+         // we need to try to wrap the entry to get it removed
+         // for the removal itself, wrapping null would suffice, but listeners need previous value
+         entryFactory.wrapEntryForWriting(ctx, key, true, false);
+         if (trace)
+           log.tracef("Entry to be removed: %s", toStr(key));
       }
-      return setSkipRemoteGetsAndInvokeNextForWriteCommand(ctx, command);
+      return setSkipRemoteGetsAndInvokeNextForManyEntriesCommand(ctx, command);
    }
 
    @Override
    public final BasicInvocationStage visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command)
          throws Throwable {
-      wrapEntryForPutIfNeeded(ctx, command);
+      wrapEntryIfNeeded(ctx, command);
       return setSkipRemoteGetsAndInvokeNextForDataCommand(ctx, command, command.getMetadata());
    }
 
-   private void wrapEntryForPutIfNeeded(InvocationContext ctx, AbstractDataWriteCommand command) throws Throwable {
-      if (shouldWrap(command.getKey(), ctx, command)) {
-         boolean skipRead = command.hasFlag(Flag.IGNORE_RETURN_VALUES) && !command.isConditional();
-         entryFactory.wrapEntryForWriting(ctx, command.getKey(), EntryFactory.Wrap.WRAP_ALL, skipRead, false);
-      }
+   private void wrapEntryIfNeeded(InvocationContext ctx, AbstractDataWriteCommand command) throws Throwable {
+      removeFromContextOnRetry(ctx, command);
+      boolean skipRead = command.hasFlag(Flag.IGNORE_RETURN_VALUES) && !command.isConditional();
+      entryFactory.wrapEntryForWriting(ctx, command.getKey(), skipRead, ignoreOwnership(command) || canRead(command.getKey()));
    }
 
-   private boolean shouldWrap(Object key, InvocationContext ctx, FlagAffectedCommand command) {
-      if (command.hasFlag(Flag.SKIP_OWNERSHIP_CHECK)) {
-         if (trace)
-            log.tracef("Skipping ownership check and wrapping key %s", toStr(key));
-
-         return true;
-      } else if (command.hasFlag(Flag.CACHE_MODE_LOCAL)) {
+   private void removeFromContextOnRetry(InvocationContext ctx, AbstractDataWriteCommand command) {
+      // When originator is a backup and it becomes primary (and we retry the command), the context already
+      // contains the value before the command started to be executed. However, another modification could happen
+      // after this node became an owner, so we have to force a reload.
+      // With repeatable reads, we cannot just remove the entry from context; instead of we will rely
+      // on the write skew check to do the reload & comparison in the end.
+      // TODO: what about pessimistic locking?
```

Comment:
> Eh, replace value in DC from the context?

I was thinking of replacing the value in the `MVCCEntry` with the previous value from the same `MVCCEntry`.

> I am not breaking RR for pessimistic tx, but a concurrent modification may be missed, therefore the conditional operation won't behave correctly. Though, if we're already acquired the lock, there shouldn't be a concurrent op. The TODO was more about reasoning about that :)
> And yes, WSC will be thrown only in case of concurrent modification.

Then I guess the comment is no longer needed? It makes it look like the implementation is incomplete.

> Are there any tests that simulate retries on tx cache?

Check out `TxReplay*Test`. Of course, they don't test retries of tx write commands before the commit.